### PR TITLE
link to boards

### PR DESCRIPTION
**Pull request by alixander**
```
x.link: layers.y
layers: {
  y: {
    home.link: _
    layers: {
      x: {
        parent.link: _._.layers.y
      }
    }
  }
}
```

gets transformed into

```
x.link: root.layers.y
layers: {
  y: {
    home.link: root
    layers: {
      x: {
        parent.link: root.layers.y
      }
    }
  }
}
```

this will make it easy for exporters to link board together in whatever format

### CODE REVIEW

File: d2compiler/compile_test.go [hunk 1]
```diff
@@ -2027,6 +2027,121 @@ Chinchillas_Collectibles.chinchilla -> Chinchillas.id`,
...
+
+layers: {
+  cat: {
+    the cat -> meeeowwww: goes
+  }
+}
+
+scenarios: {
+  green: {
+		question.style.fill: green
```
Comment (nhooyr) [inline d2compiler/compile_test.go:2055]: bad indent

File: d2compiler/compile_test.go [hunk 2]
```diff
@@ -2027,6 +2027,121 @@ Chinchillas_Collectibles.chinchilla -> Chinchillas.id`,
...
+			text: `x.link: layers.x
+`,
+			expErr: `d2/testdata/d2compiler/TestCompile/link-board-not-found.d2:1:1: linked board not found`,
+		},
+		{
+			name: "link-board-not-board",
+			text: `zzz
+x.link: layers.x.y
+layers: {
+	x: {
```
Comment (nhooyr) [inline d2compiler/compile_test.go:2074]: bad indent.

File: d2compiler/compile_test.go [hunk 3]
```diff
@@ -2027,6 +2027,121 @@ Chinchillas_Collectibles.chinchilla -> Chinchillas.id`,
...
+    y
+  }
+}`,
+			expErr: `d2/testdata/d2compiler/TestCompile/link-board-not-board.d2:2:1: linked board not found`,
+		},
+		{
+			name: "link-board-nested",
+			text: `x.link: layers.x.layers.x
+layers: {
```
Comment (nhooyr) [inline d2compiler/compile_test.go:2083]: Bad indent.

File: d2compiler/compile_test.go [hunk 4]
```diff
@@ -2027,6 +2027,121 @@ Chinchillas_Collectibles.chinchilla -> Chinchillas.id`,
...
+				tassert.Equal(t, "root.layers.x.layers.x", g.Objects[0].Attributes.Link.Value)
+			},
+		},
+		{
+			name: "link-board-key-nested",
+			text: `x: {
+  y.link: layers.x
+}
+layers: {
+	x: {
```
Comment (nhooyr) [inline d2compiler/compile_test.go:2102]: Bad indent.

File: d2compiler/compile_test.go [hunk 5]
```diff
@@ -2027,6 +2027,121 @@ Chinchillas_Collectibles.chinchilla -> Chinchillas.id`,
...
+			},
+		},
+		{
+			name: "link-board-underscore",
+			text: `x
+layers: {
+	x: {
+	  yo
+    layers: {
+			x: {
```
Comment (nhooyr) [inline d2compiler/compile_test.go:2117]: .

File: d2compiler/compile_test.go [hunk 6]
```diff
@@ -2027,6 +2027,121 @@ Chinchillas_Collectibles.chinchilla -> Chinchillas.id`,
...
+			assertions: func(t *testing.T, g *d2graph.Graph) {
+				tassert.NotNil(t, g.Layers[0].Layers[0].Objects[0].Attributes.Link.Value)
+				tassert.Equal(t, "root.layers.x", g.Layers[0].Layers[0].Objects[0].Attributes.Link.Value)
+				tassert.Equal(t, "root.layers.x", g.Layers[0].Layers[0].Objects[1].Attributes.Link.Value)
+			},
+		},
+		{
+			name: "link-board-underscore-not-found",
+			text: `x
+layers: {
```
Comment (nhooyr) [inline d2compiler/compile_test.go:2133]: .

File: d2ir/compile.go [hunk 7]
```diff
@@ -127,13 +129,74 @@ func (c *compiler) compileField(dst *Map, kp *d2ast.KeyPath, refctx *RefContext)
 		}
 		c.compileMap(f.Map(), refctx.Key.Value.Map)
 	} else if refctx.Key.Value.ScalarBox().Unbox() != nil {
+		// If these are boards, we transform into absolute paths
```
Comment (nhooyr) [inline d2ir/compile.go:132]:
```suggestion
		// If the link is a board, we need to transform it into an absolute path.
```

File: d2ir/d2ir.go [hunk 8]
```diff
@@ -221,6 +221,17 @@ func (m *Map) Root() bool {
 	return f.Root()
 }
 
+func (m *Map) AbsID() string {
```
Comment (nhooyr) [inline d2ir/d2ir.go:224]: See global `IDA()` function. Copy same code of global IDA and remove `|| NodeBoardKind(f) != ""` on the 4th line and name it `AbsIDA()` and make it a global too.

File: d2ir/compile.go [hunk 9]
```diff
@@ -127,13 +129,74 @@ func (c *compiler) compileField(dst *Map, kp *d2ast.KeyPath, refctx *RefContext)
...
+	if len(scopeIDA) == 0 {
+		return
+	}
+
+	linkIDA := link.IDA()
+	if len(linkIDA) == 0 {
+		return
+	}
+
+	// If it doesn't start with one of these reserved words, the link may be a URL or local path or something
```
Comment (nhooyr) [inline d2ir/compile.go:162]:
```suggestion
	// If it doesn't start with one of these reserved words, the link is definitely not a board link.
```

File: d2ir/compile.go [hunk 10]
```diff
@@ -127,13 +129,74 @@ func (c *compiler) compileField(dst *Map, kp *d2ast.KeyPath, refctx *RefContext)
...
+	if linkIDA[0] != "layers" && linkIDA[0] != "scenarios" && linkIDA[0] != "steps" && linkIDA[0] != "_" {
+		return
+	}
+
+	// Chop off the non-board portion of the scope, like if this is being defined on a nested object (e.g. `x.y.z`)
+	for i := len(scopeIDA) - 1; i > 0; i-- {
```
Comment (nhooyr) [inline d2ir/compile.go:168]:
```suggestion
	for i := len(scopeIDA) - 1; i > 1; i-- {
```

File: d2ir/compile.go [hunk 11]
```diff
@@ -127,13 +129,74 @@ func (c *compiler) compileField(dst *Map, kp *d2ast.KeyPath, refctx *RefContext)
...
+		if scopeIDA[i-1] == "layers" || scopeIDA[i-1] == "scenarios" || scopeIDA[i-1] == "steps" {
+			scopeIDA = scopeIDA[:i+1]
+			break
+		}
+		if scopeIDA[i-1] == "root" {
```
Comment (nhooyr) [inline d2ir/compile.go:173]: So we do have a way now to refer to the root board and it's not considered an error if a user inputs it. I still disagree but I'll defer.

File: d2ir/compile.go [hunk 12]
```diff
@@ -127,13 +129,74 @@ func (c *compiler) compileField(dst *Map, kp *d2ast.KeyPath, refctx *RefContext)
...
+			scopeIDA = scopeIDA[:i]
+			break
+		}
+	}
+
+	// Resolve underscores
+	for len(linkIDA) > 0 && linkIDA[0] == "_" {
+		if len(scopeIDA) < 2 {
+			// IR compiler only validates bad underscore usage
```
Comment (nhooyr) [inline d2ir/compile.go:182]: Why not make this `"invalid underscore usage"` then?

File: d2ir/compile.go [hunk 13]
```diff
@@ -127,13 +129,74 @@ func (c *compiler) compileField(dst *Map, kp *d2ast.KeyPath, refctx *RefContext)
...
+		scopeIDA = scopeIDA[:len(scopeIDA)-2]
+		linkIDA = linkIDA[1:]
+	}
+	if len(scopeIDA) == 0 {
+		scopeIDA = []string{"root"}
+	}
+
+	// Create the absolute path by appending scope path with value specified
+	scopeIDA = append(scopeIDA, linkIDA...)
+	refctx.Key.Value = d2ast.MakeValueBox(d2ast.RawString(strings.Join(scopeIDA, "."), true))
```
Comment (nhooyr) [inline d2ir/compile.go:197]: This won't work robustly. You'll get parsing errors later if special characters are used in the name of boards. You need to convert the IDA into a *d2ast.KeyPath and then use d2format.Format that into a string that you pass into RawString.

**Review by nhooyr — CHANGES_REQUESTED** [review]
not as long as I expected.

File: d2compiler/compile.go [hunk 14]
```diff
@@ -715,6 +720,69 @@ func (c *compiler) validateNear(g *d2graph.Graph) {
...
+		if err != nil {
+			continue
+		}
+
+		if linkKey.Path[0].Unbox().ScalarString() != "root" {
+			continue
+		}
+
+		root := g
+		for root.Parent != nil {
```
[the remark this answers is not in the document]
Comment (alixander) [reply d2compiler/compile.go:739]: there's arleady a field named Root and i don't wanna do that `Root_` rename

File: d2ir/d2ir.go (hunk 8 above)
Comment (nhooyr) [reply d2ir/d2ir.go:224]: Or actually rename the global IDA to `BoardIDA()` and make this new one just `IDA()`

Comment (nhooyr) [reply d2ir/d2ir.go:224]: And then delete this of course and use the global.

File: d2compiler/compile.go [hunk 15]
```diff
@@ -715,6 +720,69 @@ func (c *compiler) validateNear(g *d2graph.Graph) {
...
+	for _, b := range append(append(g.Layers, g.Scenarios...), g.Steps...) {
+		c.validateBoardLinks(b)
+	}
+}
+
+func hasBoard(root *d2graph.Graph, ida []string) bool {
+	for i := 0; i < len(ida); i += 2 {
+		id := ida[i]
+		if id == "root" {
+			i--
```
[the remark this answers is not in the document]
Comment (alixander) [reply d2compiler/compile.go:756]: root is a special case.

look at
`root.layers.x.layers.y`

the pattern is always `type.name.type.name`, except root, which is `type.type.name.type.name`

File: d2compiler/compile.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (nhooyr) [reply d2compiler/compile.go:739]: Just call it `RootBoard()`.

File: d2compiler/compile.go [hunk 16]
```diff
@@ -715,6 +720,69 @@ func (c *compiler) validateNear(g *d2graph.Graph) {
...
+					return hasBoard(b, ida[i+1:])
+				}
+			}
+		} else if id == "steps" {
+			for _, b := range root.Steps {
+				if b.Name == nextID {
+					return hasBoard(b, ida[i+1:])
+				}
+			}
+		}
```
[the remark this answers is not in the document]
Comment (alixander) [reply d2compiler/compile.go:781]: oh damn, good catch

File: d2compiler/compile.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (nhooyr) [reply d2compiler/compile.go:756]: Ah right yea this is fine then.

[the remark this answers is not in the document]
Comment (nhooyr) [reply d2compiler/compile.go:756]: Though perhaps you should check if `root` is the very first? Here you're allowing it anywhere.

File: d2ir/d2ir.go (hunk 8 above)
Comment (alixander) [reply d2ir/d2ir.go:224]: ah. missed that function

File: d2ir/compile.go (hunk 10 above)
Comment (alixander) [reply d2ir/compile.go:168]: `link-board-key-nested` test fails with this change:

```
x: {
  y.link: layers.x
}
layers: {
  x: {
    yo
  }
}
```

the scope of `x.y.link` should be `root`. this loop needs to go all the way to 0th, and see `root`, and chop everything after

File: d2ir/compile.go (hunk 11 above)
Comment (alixander) [reply d2ir/compile.go:173]: oh it was unintentional. i suppose the code as is would work for absolute paths though.

File: d2ir/compile.go (hunk 10 above)
Comment (nhooyr) [reply d2ir/compile.go:168]: Oh right, sorry I misunderstood this loop, I thought it was >= 0, thought it'd panic at 0 but nah it's fine.

File: d2ir/compile.go (hunk 11 above)
Comment (nhooyr) [reply d2ir/compile.go:173]: I'd prefer you make it an error then if it was unintentional.

File: d2ir/compile.go (hunk 13 above)
Comment (alixander) [reply d2ir/compile.go]: like this?

https://github.com/terrastruct/d2/pull/766/commits/78e6f3fc06f4d8292d889e8d1ef86d16ecbcd167#diff-d1f402f2e19396effb60a83e703aa9de4ff768e4a2fe1dd5b7f6b4a0da47667cR195

File: d2ir/compile.go (hunk 11 above)
Comment (alixander) [reply d2ir/compile.go:176]: ok

File: d2format/format.go [hunk 17]
```diff
@@ -409,3 +409,11 @@ func KeyPath(kp *d2ast.KeyPath) (ida []string) {
 	}
 	return ida
 }
+
+func IDA(ida []string) *d2ast.KeyPath {
```
Comment (nhooyr) [inline d2format/format.go:413]: Nice though perhaps this should be in d2ast instead?

**Review by nhooyr — CHANGES_REQUESTED** [review]

File: d2compiler/compile.go [hunk 18]
```diff
@@ -735,49 +735,54 @@ func (c *compiler) validateBoardLinks(g *d2graph.Graph) {
...
 			}
-		} else if id == "steps" {
+		case "steps":
 			for _, b := range root.Steps {
 				if b.Name == nextID {
 					return hasBoard(b, ida[i+1:])
 				}
 			}
+		default:
+			break
```
Comment (nhooyr) [inline d2compiler/compile.go:785]: This will break the switch not the loop. Add a loop label.

File: d2format/format.go (hunk 17 above)
Comment (alixander) [reply d2format/format.go:413]: yeah want me to get the one above this too? it's just a mirror

File: d2ir/d2ir.go [hunk 19]
```diff
@@ -1064,6 +1053,26 @@ func IDA(n Node) (ida []string) {
 	}
 }
 
+// IDA returns the absolute path to n from the nearest board root.
```
Comment (nhooyr) [inline d2ir/d2ir.go:1056]: Cut off the comment at `absolute path to n.`

**Review by nhooyr — CHANGES_REQUESTED** [review]
🚀

File: d2format/format.go (hunk 17 above)
Comment (nhooyr) [reply d2format/format.go:413]: Nah, that is d2format specific. Oh lmfao we already have IDA in d2ast, it's called `d2ast.MakeKeyPath`.

Comment (alixander) [reply d2format/format.go]: damnit i thought there might be

File: d2compiler/compile.go [hunk 20]
```diff
@@ -781,9 +781,8 @@ func hasBoard(root *d2graph.Graph, ida []string) bool {
 					return hasBoard(b, ida[i+1:])
 				}
 			}
-		default:
-			break
 		}
+		break
```
Comment (nhooyr) [inline d2compiler/compile.go:785]: ✅

**Review by nhooyr — APPROVED** [review]
🚀

**Pull request merged**